avm/src/vm/memory.rs

```diff
@@ -204,174 +202,143 @@ impl HandlerMemory {
+      block,
+      hm_id: Arc::as_ptr(hm) as usize,
+    }, is_fractal);
   }
 
   /// Simply sets a given address to an explicit set of `mems` indexes. Simplifies pointer creation
   /// to deeply-nested data.
-  fn set_addr(self: &mut HandlerMemory, addr: i64, a: usize, b: usize) {
+  fn set_addr(self: &mut Arc<HandlerMemory>, addr: i64, a: usize, b: usize) {
+    let hm = Arc::get_mut(self).expect("unable to set address in HM: dangling pointer");
     if addr_type(addr) == NORMAL_ADDR {
       let addru = addr as usize;
-      if self.addr.0.len() <= addru {
-        self.addr.0.resize(addru + 1, None);
+      if hm.addr.0.len() <= addru {
+        hm.addr.0.resize(addru + 1, None);
       }
-      self.addr.0[addru] = Some((a, b));
+      hm.addr.0[addru] = Some((a, b));
     } else {
       let addru = (addr - CLOSURE_ARG_MEM_START) as usize;
-      if self.addr.1.len() <= addru {
-        self.addr.1.resize(addru + 1, None);
+      if hm.addr.1.len() <= addru {
+        hm.addr.1.resize(addru + 1, None);
       }
-      self.addr.1[addru] = Some((a, b));
+      hm.addr.1[addru] = Some((a, b));
     }
   }
 
   /// For the memory block(s) starting at idx in Fractal, determines if the data is a single value or an array of
   /// values, and returns that value either as a vector or the singular value wrapped in a vector,
   /// and a boolean indicating if it is a fractal value or not.
   pub fn read_from_fractal(
-    self: &HandlerMemory,
+    self: &Arc<HandlerMemory>,
     fractal: &FractalMemory,
     idx: usize,
   ) -> (FractalMemory, bool) {
     let (a, b) = fractal.block[idx];
     let b_usize = b as usize;
-    let hm_opt = self.hm_for_idxs(a, b as usize);
-    let hm = if hm_opt.is_none() {
-      self
-    } else {
-      hm_opt.unwrap()
+    let hm_opt = self.hm_for_idxs(a, b_usize);
+    let hm = match hm_opt.as_ref() {
+      Some(hm) => hm,
+      None => self,
     };
     return if a == std::usize::MAX {
       // The indexes are the actual data
       (FractalMemory::new(vec![(a, b)]), false)
-    } else if b_usize < std::usize::MAX {
-      // The indexes point to fixed data
-      (
-        FractalMemory {
-          hm_addr: None,
-          block: vec![hm.mems[a][b_usize].clone()],
-          hm_id: hm as *const HandlerMemory as usize,
-        },
-        false,
-      )
     } else {
-      // The indexes point at nested data
-      (
-        FractalMemory {
-          hm_addr: None,
-          block: hm.mems[a].clone(),
-          hm_id: hm as *const HandlerMemory as usize,
-        },
-        true,
-      )
+      let (block, is_fractal) = if b_usize < std::usize::MAX {
+        // The indexes point to fixed data
+        (vec![hm.mems[a][b_usize].clone()], false)
+      } else { // b_usize > std::usize::MAX
+        // The indexes point at nested data
+        (hm.mems[a].clone(), true)
+      };
+      (FractalMemory {
+        hm_addr: None,
+        block,
+        hm_id: Arc::as_ptr(hm) as usize,
```

Comment:
I remember you pointed out this doesn't actually work like I thought it did when I implemented. Does it work as a unique id with Arcs? Maybe we should remove all together

Reply:
I'm doing the same thing as you did before, but the difference is that the references you were using before were references to the object on the stack. I'm not 100% sure how often Rust allows for moving values on the stack, but it's entirely possible for the `HandlerMemory` to not be at the same memory address between the time the `FractalMemory` is created and the time its `hm_id` field is used